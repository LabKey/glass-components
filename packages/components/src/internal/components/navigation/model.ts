/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { List, Record } from 'immutable';
import { Ajax, Utils, QueryKey } from '@labkey/api';

import { buildURL, createProductUrl, AppURL, createProductUrlFromPartsWithContainer } from '../../url/AppURL';

export class MenuSectionModel extends Record({
    label: undefined,
    url: undefined,
    items: List<MenuItemModel>(),
    totalCount: 0,
    key: undefined,
    productId: undefined,
    sectionKey: undefined,
}) {
    declare label: string;
    declare url: string;
    declare items: List<MenuItemModel>;
    declare totalCount: number;
    declare key: string;
    declare productId: string;
    declare sectionKey: string;

    static create(rawData: any, currentProductId?: string, containerPath?: string): MenuSectionModel {
        if (rawData) {
            let items;

            if (rawData.items) {
                items = rawData.items.map(i =>
                    MenuItemModel.create(i, rawData.sectionKey, currentProductId, containerPath)
                );
            }

            return new MenuSectionModel(
                Object.assign({}, rawData, {
                    items: List(items),
                })
            );
        }

        return new MenuSectionModel();
    }

    getVisibleItems(): MenuItemModel[] {
        return this.items.filter(item => !item.hidden).toArray();
    }

    hasVisibleItems(): boolean {
        return this.getVisibleItems().length > 0;
    }
}

export class MenuItemModel extends Record({
    id: undefined,
    key: undefined,
    label: undefined,
    url: undefined,
    hidden: false,
    orderNum: undefined,
    originalUrl: undefined,
    requiresLogin: false,
    hasActiveJob: false,
    fromSharedContainer: false,
}) {
    declare id: number;
    declare key: string;
    declare label: string;
    declare url: string | AppURL;
    declare hidden: boolean;
    declare orderNum: number;
    declare originalUrl: string;
    declare requiresLogin: boolean;
    declare hasActiveJob: boolean;
    declare fromSharedContainer: boolean;

    static create(rawData, sectionKey: string, currentProductId?: string, containerPath?: string): MenuItemModel {
        if (rawData) {
            const dataProductId = rawData.productId ? rawData.productId.toLowerCase() : undefined;

            if (rawData.key && sectionKey !== 'user') {
                // for assay names and freezer names that contain slashes, full raw key (protocol/assayname: general/a/b) is encoded using QueryKey.encodePart as general/a$Sb on server side
                // use QueryKey.decodePart to decode the name so url can be correctly called by encodeURIComponent and key be used to display decoded assay name
                const subParts = rawData.key
                    .split('/')
                    .filter(val => val !== '')
                    .map(QueryKey.decodePart);

                const decoded = subParts.join('/');
                const decodedKey = rawData.key.replace(rawData.key, () => decoded); // use the functional version to skip any additional pattern substitutions https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace#specifying_a_string_as_the_replacement

                return new MenuItemModel(
                    Object.assign({}, rawData, {
                        originalUrl: rawData.url,
                        url: createProductUrlFromPartsWithContainer(
                            dataProductId,
                            currentProductId,
                            containerPath,
                            undefined,
                            sectionKey,
                            ...subParts
                        ),
                        key: decodedKey,
                    })
                );
            } else {
                return new MenuItemModel(
                    Object.assign({}, rawData, {
                        originalUrl: rawData.url,
                        url: createProductUrl(dataProductId, currentProductId, rawData.url, containerPath),
                    })
                );
            }
        }
        return new MenuItemModel();
    }

    getUrlString(useOriginalUrl?: boolean): string {
        if (useOriginalUrl && this.originalUrl) return this.originalUrl;
        return typeof this.url === 'string' ? this.url : this.url?.toHref();
    }
}

export class ProductMenuModel extends Record({
    isError: false,
    isLoaded: false,
    isLoading: false,
    sections: List<MenuSectionModel>(),
    message: undefined,
    containerId: undefined,
    containerPath: undefined,
    currentProductId: undefined,
    productIds: undefined,
    needsReload: false,
}) {
    declare isError: boolean;
    declare isLoaded: boolean;
    declare isLoading: boolean;
    declare message: string;
    declare sections: List<MenuSectionModel>;
    declare containerId: string;
    declare containerPath: string;
    declare currentProductId: string; // the current product's id
    declare productIds: List<string>; // the list of all product ids to be included in the menu; leave undefined for all products in the container
    declare needsReload: boolean;

    /**
     * Retrieve the product menu sections for this productId
     */
    getMenuSections(): Promise<List<MenuSectionModel>> {
        return new Promise((resolve, reject) => {
            return Ajax.request({
                url: buildURL('product', 'menuSections.api', undefined, {
                    container: this.containerPath,
                }),
                params: Object.assign({
                    currentProductId: this.currentProductId,
                    productIds: List.isList(this.productIds) ? this.productIds.toArray().join(',') : this.productIds,
                }),
                success: Utils.getCallbackWrapper(response => {
                    let sections = List<MenuSectionModel>();
                    if (response) {
                        response.forEach(sectionData => {
                            sections = sections.push(
                                MenuSectionModel.create(sectionData, this.currentProductId, this.containerPath)
                            );
                        });
                    }
                    resolve(sections);
                }),
                failure: Utils.getCallbackWrapper(response => {
                    console.error(response);
                    reject(response);
                }),
            });
        });
    }

    setLoadedSections(sections: List<MenuSectionModel>): ProductMenuModel {
        return this.merge({
            isLoaded: true,
            isLoading: false,
            needsReload: false,
            sections,
        }) as ProductMenuModel;
    }

    setError(message: string): ProductMenuModel {
        return this.merge({
            isLoaded: true,
            isLoading: false,
            needsReload: false,
            isError: true,
            message,
        }) as ProductMenuModel;
    }

    getSection(key: string): MenuSectionModel {
        return this.sections.find(section => section.key.toLowerCase() === key.toLowerCase());
    }

    hasSectionItems(key: string): boolean {
        return this.isLoaded && this.getSection(key)?.totalCount > 0;
    }

    setNeedsReload(): ProductMenuModel {
        return this.merge({
            needsReload: true,
        }) as ProductMenuModel;
    }
}

export class MenuSectionConfig extends Record({
    activeJobIconCls: 'fa-spinner fa-pulse',
    emptyText: undefined,
    filteredEmptyText: undefined,
    emptyAppURL: undefined,
    emptyURLText: 'Get started...',
    headerURLPart: undefined,
    headerText: undefined,
    iconCls: undefined,
    iconURL: undefined,
    showActiveJobIcon: true,
    staticContent: false,
    useOriginalURL: false,
}) {
    declare activeJobIconCls?: string;
    declare emptyText?: string;
    declare filteredEmptyText?: string;
    declare emptyAppURL?: AppURL;
    declare emptyURLText: string;
    declare headerURLPart: string;
    declare headerText?: string;
    declare iconCls?: string;
    declare iconURL?: string;
    declare showActiveJobIcon?: boolean;
    // Inform the display that this section's content is static (unchanging).
    // This helps inform the layout when these sections are laid out alongside sections with dynamic content.
    declare staticContent?: boolean;
    declare useOriginalURL?: boolean;
}
