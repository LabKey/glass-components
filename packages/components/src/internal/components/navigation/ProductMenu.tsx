/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React, { MouseEvent, FC, memo, useCallback, useState, useEffect, useRef, useMemo } from 'react';
import classNames from 'classnames';
import { List, Map } from 'immutable';
import { DropdownButton } from 'react-bootstrap';
import { withRouter, WithRouterProps } from 'react-router';
import { ActionURL } from '@labkey/api';

import { blurActiveElement } from '../../util/utils';
import { LoadingSpinner } from '../base/LoadingSpinner';
import { useServerContext } from '../base/ServerContext';
import { AppProperties } from '../../app/models';
import { getCurrentAppProperties } from '../../app/utils';

import { Alert } from '../base/Alert';

import { isLoading, LoadingState } from '../../../public/LoadingState';
import { naturalSortByProperty } from '../../../public/sort';
import { resolveErrorMessage } from '../../util/messaging';
import { AppContext, useAppContext } from '../../AppContext';
import { Container } from '../base/models/Container';
import { buildURL } from '../../url/AppURL';

import {
    AUDIT_KEY,
    MEDIA_KEY,
    REGISTRY_KEY,
    SAMPLE_TYPE_KEY,
    SAMPLES_KEY,
    SEARCH_KEY,
    WORKFLOW_KEY,
    ASSAY_DESIGN_KEY,
    ASSAYS_KEY,
    PICKLIST_KEY,
    ELN_KEY,
    SOURCE_TYPE_KEY,
    SOURCES_KEY,
    FREEZERS_KEY,
    BOXES_KEY,
} from '../../app/constants';

import { FolderMenu, FolderMenuItem } from './FolderMenu';
import { ProductMenuSection } from './ProductMenuSection';
import { MenuSectionConfig, MenuSectionModel, ProductMenuModel } from './model';
import { initMenuModel } from './actions';

interface ProductMenuButtonProps {
    appProperties?: AppProperties;
    sectionConfigs: List<Map<string, MenuSectionConfig>>;
    showFolderMenu: boolean;
}

const ProductMenuButtonImpl: FC<ProductMenuButtonProps & WithRouterProps> = memo(props => {
    const { appProperties = getCurrentAppProperties(), routes } = props;
    const [menuOpen, setMenuOpen] = useState(false);
    const [error, setError] = useState<string>();
    const [loading, setLoading] = useState<LoadingState>(LoadingState.INITIALIZED);
    const [folderItems, setFolderItems] = useState<FolderMenuItem[]>([]);
    const hasError = !!error;
    const isLoaded = !isLoading(loading);
    const { api } = useAppContext<AppContext>();
    const { container } = useServerContext();

    useEffect(() => {
        setLoading(LoadingState.LOADING);
        setError(undefined);

        (async () => {
            try {
                const folders = await api.security.fetchContainers({
                    // Container metadata does not always provide "type" so inspecting the
                    // "parentPath" to determine top-level folder vs subfolder.
                    containerPath: container.parentPath === '/' ? container.path : container.parentPath,
                });

                const items_: FolderMenuItem[] = [];
                const topLevelFolderIdx = folders.findIndex(f => f.parentPath === '/');
                if (topLevelFolderIdx > -1) {
                    // Remove top-level folder from array as it is always displayed as the first menu item
                    const topLevelFolder = folders.splice(topLevelFolderIdx, 1)[0];
                    items_.push(createFolderItem(topLevelFolder, appProperties.controllerName, true));
                }

                // Issue 45805: sort folders by title as server-side sorting is insufficient
                folders.sort(naturalSortByProperty('title'));
                setFolderItems(
                    items_.concat(folders.map(folder => createFolderItem(folder, appProperties.controllerName, false)))
                );
            } catch (e) {
                setError(`Error: ${resolveErrorMessage(e)}`);
            }

            setLoading(LoadingState.LOADED);
        })();
    }, [api, container, appProperties?.controllerName]);

    const toggleMenu = useCallback(() => {
        setMenuOpen(!menuOpen);
        blurActiveElement();
    }, [menuOpen, setMenuOpen]);

    // Only toggle the menu closing if a menu section link has been clicked.
    // Clicking anywhere else inside the menu will not toggle the menu, including side panel folder clicks.
    const onClick = useCallback(
        (evt: MouseEvent<HTMLDivElement>) => {
            const { nodeName, className } = evt.target as any;
            if (!nodeName || (nodeName.toLowerCase() === 'a' && className !== 'menu-folder-item')) {
                toggleMenu();
            }
        },
        [toggleMenu]
    );

    const subtitle = useMemo(() => {
        return getHeaderMenuSubtitle(routes?.[1]?.path);
    }, [routes]);

    if (!isLoaded && !hasError) return null;
    const showFolders = folderItems?.length > 1;
    const title = showFolders ? container.title : 'Menu';

    return (
        <DropdownButton
            className="product-menu-button"
            id="product-menu"
            onToggle={toggleMenu}
            open={menuOpen}
            title={
                <>
                    <div className="title">{title}</div>
                    <div className="subtitle">{subtitle}</div>
                </>
            }
        >
            {menuOpen && (
                <ProductMenu
                    {...props}
                    className={classNames({ 'with-col-folders': showFolders })}
                    onClick={onClick}
                    error={error}
                    folderItems={folderItems}
                    showFolderMenu={showFolders}
                />
            )}
        </DropdownButton>
    );
});

export const ProductMenuButton = withRouter<ProductMenuButtonProps>(ProductMenuButtonImpl);

interface ProductMenuProps extends ProductMenuButtonProps {
    className: string;
    error: string;
    folderItems: FolderMenuItem[];
    onClick: (evt: MouseEvent<HTMLDivElement>) => void;
}

export const ProductMenu: FC<ProductMenuProps> = memo(props => {
    const {
        className,
        onClick,
        error,
        folderItems,
        sectionConfigs,
        showFolderMenu,
        appProperties = getCurrentAppProperties(),
    } = props;
    const { container, moduleContext } = useServerContext();
    const [menuModel, setMenuModel] = useState<ProductMenuModel>(new ProductMenuModel({ containerId: container.id }));
    const contentRef = useRef<HTMLDivElement>();

    useEffect(() => {
        if (!menuModel.isLoaded) return;
        let height = 400; // match navbar.scss product-menu-content min-height
        const maxHeight = window.innerHeight * 0.8;
        const sections = Array.from(contentRef.current.getElementsByClassName('menu-section'));
        sections.forEach(section => (height = Math.max(height, section.clientHeight)));
        contentRef.current.style.height = Math.min(height, maxHeight) + 'px';

        // if the selected project is out of view, scrollIntoView
        const activeProject = contentRef.current.getElementsByClassName('active')?.[0];
        if (activeProject) {
            if (activeProject.getBoundingClientRect().bottom > contentRef.current.getBoundingClientRect().bottom) {
                contentRef.current.getElementsByClassName('active')?.[0].scrollIntoView({ behavior: 'smooth' });
            }
        }
    }, [menuModel.isLoaded]);

    useEffect(() => {
        (async () => {
            // no try/catch as the initMenuModel will catch errors and put them in the model isError/message
            const menuModel_ = await initMenuModel(appProperties, moduleContext, container.id);
            setMenuModel(menuModel_);
        })();
    }, [appProperties, container.id, moduleContext]);

    const onFolderItemClick = useCallback(
        async (folderItem: FolderMenuItem) => {
            // return early if folderItem is already active
            if (folderItem.id === menuModel.containerId) return;

            setMenuModel(new ProductMenuModel({ containerId: folderItem.id })); // loading state, reset error

            // no try/catch as the initMenuModel will catch errors and put them in the model isError/message
            const containerPath = folderItem.id === container.id ? undefined : folderItem.path;
            const menuModel_ = await initMenuModel(appProperties, moduleContext, folderItem.id, containerPath);
            setMenuModel(menuModel_);
        },
        [appProperties, container.id, menuModel.containerId, moduleContext]
    );

    const getSectionModel = useCallback(
        (key: string): MenuSectionModel => menuModel.sections.find(section => section.key === key),
        [menuModel]
    );

    const dashboardURL = useMemo(() => {
        return showFolderMenu && appProperties.logoBadgeColorImageUrl;
    }, [appProperties.logoBadgeColorImageUrl, showFolderMenu]);

    return (
        <div className={classNames('product-menu-content', className)} onClick={onClick} ref={contentRef}>
            <div className="navbar-connector" />
            {error && <Alert>{error}</Alert>}
            {showFolderMenu && (
                <FolderMenu activeContainerId={menuModel.containerId} items={folderItems} onClick={onFolderItemClick} />
            )}
            <div className="sections-content">
                {!menuModel.isLoaded && (
                    <div className="menu-section menu-loading">
                        <LoadingSpinner />
                    </div>
                )}
                {menuModel.isError && (
                    <div className="menu-section">
                        <Alert>{menuModel.message}</Alert>
                    </div>
                )}
                {menuModel.isLoaded &&
                    sectionConfigs.map((sectionConfig, i) => (
                        <div key={i} className="menu-section col-product-section">
                            {sectionConfig.entrySeq().map(([key, menuConfig], j) => {
                                const isLast = i === sectionConfigs.size - 1 && j === sectionConfig.size - 1;

                                return (
                                    <ProductMenuSection
                                        key={key}
                                        section={getSectionModel(key)}
                                        config={menuConfig}
                                        containerPath={menuModel.containerPath}
                                        currentProductId={menuModel.currentProductId}
                                        dashboardImgURL={isLast && dashboardURL}
                                    />
                                );
                            })}
                        </div>
                    ))}
            </div>
        </div>
    );
});

function createFolderItem(folder: Container, controllerName: string, isTopLevel: boolean): FolderMenuItem {
    return {
        href: buildURL(controllerName, `${ActionURL.getAction() || 'app'}.view`, undefined, {
            container: folder.path,
            returnUrl: false,
        }),
        id: folder.id,
        isTopLevel,
        label: folder.title,
        path: folder.path,
    };
}

const HEADER_MENU_SUBTITLE_MAP = {
    account: 'Settings',
    admin: 'Administration',
    items: 'Storage',
    lineage: 'Lineage',
    home: 'Dashboard',
    pipeline: 'Imports',
    q: 'Schemas',
    reports: 'Reports',

    [ASSAY_DESIGN_KEY]: 'Assays',
    [ASSAYS_KEY]: 'Assays',
    [AUDIT_KEY]: 'Administration',
    [BOXES_KEY]: 'Storage',
    [ELN_KEY]: 'Notebooks',
    [FREEZERS_KEY]: 'Storage',
    [MEDIA_KEY]: 'Media',
    [PICKLIST_KEY]: 'Picklists',
    [REGISTRY_KEY]: 'Registry',
    [SAMPLE_TYPE_KEY]: 'Sample Types',
    [SAMPLES_KEY]: 'Sample Types',
    [SOURCE_TYPE_KEY]: 'Source Types',
    [SOURCES_KEY]: 'Source Types',
    [SEARCH_KEY]: 'Search',
    [WORKFLOW_KEY]: 'Workflow',
};

function getHeaderMenuSubtitle(baseRoute: string) {
    return HEADER_MENU_SUBTITLE_MAP[baseRoute] ?? 'Dashboard';
}
