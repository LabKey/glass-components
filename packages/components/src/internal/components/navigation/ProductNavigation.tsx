import React, { FC, memo, useState } from 'react';
import { DropdownButton } from 'react-bootstrap';

import { ProductNavigationMenu } from "./ProductNavigationMenu";

interface ProductNavigationProps {

}

export const ProductNavigation: FC<ProductNavigationProps> = memo(props => {
    const [show, setShow] = useState<boolean>(false);

    return (
        <DropdownButton
            id="product-navigation-button"
            className="navbar-icon-button-right"
            noCaret={true}
            title={<i className="fa fa-th-large navbar-header-icon"/>}
            open={show}
            onToggle={() => setShow(!show)}
            pullRight={true}
        >
            <ProductNavigationMenu/>
        </DropdownButton>
    );
});
