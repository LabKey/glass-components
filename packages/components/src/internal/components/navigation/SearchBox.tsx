/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React, { ChangeEvent, FC, FormEvent, memo, useCallback, useState } from 'react';
import { FindAndSearchDropdown } from './FindAndSearchDropdown';

interface Props {
    onSearch: (value: string) => void;
    placeholder?: string;
    onFindByIds?: () => void;
    findNounPlural?: string;
}

export const SearchBox: FC<Props> = memo(props => {
    const { onSearch, placeholder, onFindByIds, findNounPlural } = props;

    const [searchValue, setSearchValue] = useState('');

    const showFindByIds = !!onFindByIds;

    const onChange = useCallback(
        (evt: ChangeEvent<HTMLInputElement>) => {
            setSearchValue(evt.target.value);
        },
        [setSearchValue]
    );

    const onSubmit = useCallback(
        (evt: FormEvent<HTMLFormElement>) => {
            evt.preventDefault();
            onSearch(searchValue);

            // reset the input value after it is has submitted
            setSearchValue('');
        },
        [onSearch, searchValue]
    );

    return (
        <form className="navbar__search-form" onSubmit={onSubmit}>
            <div className="form-group">
                <i className="fa fa-search navbar__search-icon" />
                <span className={"navbar__input-group " + (showFindByIds ? 'input-group': '')}>
                    <input
                        className="form-control navbar__search-input"
                        onChange={onChange}
                        placeholder={placeholder ?? 'Enter Search Terms'}
                        size={34}
                        type="text"
                        value={searchValue}
                    />
                    {showFindByIds &&
                    <span className="input-group-btn">
                        <FindAndSearchDropdown title={''} onFindByIds={onFindByIds} findNounPlural={findNounPlural}/>
                    </span>
                    }
                </span>
            </div>
        </form>
    );
});
