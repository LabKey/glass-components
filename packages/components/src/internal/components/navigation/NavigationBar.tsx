/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React, { FC, memo, ReactNode, useCallback } from 'react';
import { List, Map } from 'immutable';

import { User } from '../../..';

import { ServerNotifications } from '../notifications/ServerNotifications';
import { ServerNotificationsConfig } from '../notifications/model';

import { ProductNavigation } from '../productnavigation/ProductNavigation';

import { shouldShowProductNavigation } from '../productnavigation/utils';

import { ProductMenu } from './ProductMenu';
import { SearchBox } from './SearchBox';
import { UserMenu, UserMenuProps } from './UserMenu';
import { MenuSectionConfig } from './ProductMenuSection';
import { ProductMenuModel } from './model';
import { FindAndSearchDropdown } from './FindAndSearchDropdown';

interface NavigationBarProps {
    brand?: ReactNode;
    menuSectionConfigs?: List<Map<string, MenuSectionConfig>>;
    model: ProductMenuModel;
    notificationsConfig?: ServerNotificationsConfig;
    onSearch?: (form: any) => void;
    onFindByIds?: (sessionkey: string) => void;
    projectName?: string;
    searchPlaceholder?: string;
    showNavMenu?: boolean;
    showNotifications?: boolean;
    showProductNav?: boolean;
    showSearchBox?: boolean;
    user?: User;
}

type Props = NavigationBarProps & UserMenuProps;

export const NavigationBar: FC<Props> = memo(props => {
    const {
        brand,
        extraDevItems,
        extraUserItems,
        menuSectionConfigs,
        model,
        notificationsConfig,
        onSearch,
        onFindByIds,
        onSignIn,
        onSignOut,
        projectName,
        searchPlaceholder,
        showNavMenu,
        showNotifications,
        showProductNav,
        showSearchBox,
        signOutUrl,
        user,
    } = props;

    const onSearchIconClick = useCallback(() => {
        onSearch('');
    }, [onSearch]);

    const _showNotifications = showNotifications !== false && !!notificationsConfig && user && !user.isGuest;
    const _showProductNav = shouldShowProductNavigation(user) && showProductNav !== false;

    return (
        <nav className="navbar navbar-container test-loc-nav-header">
            <div className="container">
                <div className="row">
                    <div className="navbar-left col-md-5 col-sm-4 col-xs-7">
                        <span className="navbar-item pull-left">{brand}</span>
                        <span className="navbar-item-padded">
                            {showNavMenu && !!model && (
                                <ProductMenu model={model} sectionConfigs={menuSectionConfigs} />
                            )}
                        </span>
                        {projectName && (
                            <span className="navbar-item hidden-sm hidden-xs">
                                <span className="project-name">
                                    <i className="fa fa-folder-open-o" /> {projectName}{' '}
                                </span>
                            </span>
                        )}
                    </div>
                    <div className="navbar-right col-md-7 col-sm-8 col-xs-5">
                        {!!user && (
                            <div className="navbar-item pull-right">
                                <UserMenu
                                    extraDevItems={extraDevItems}
                                    extraUserItems={extraUserItems}
                                    model={model}
                                    onSignIn={onSignIn}
                                    onSignOut={onSignOut}
                                    signOutUrl={signOutUrl}
                                    user={user}
                                />
                            </div>
                        )}
                        {_showNotifications && (
                            <div className="navbar-item pull-right navbar-item-notification">
                                <ServerNotifications {...notificationsConfig} />
                            </div>
                        )}
                        {_showProductNav && (
                            <div className="navbar-item pull-right navbar-item-product-navigation">
                                <ProductNavigation />
                            </div>
                        )}
                        {showSearchBox && (
                            <div className="navbar-item pull-right">
                                <div className="hidden-sm hidden-xs">
                                    <SearchBox
                                        onSearch={onSearch}
                                        placeholder={searchPlaceholder}
                                        onFindByIds={onFindByIds}
                                        findNounPlural="samples"
                                    />
                                </div>
                                <div className="visible-sm visible-xs">
                                    {onFindByIds ? (
                                        <FindAndSearchDropdown
                                            className="navbar__xs-find-dropdown"
                                            title={<i className="fa fa-search navbar__xs-search-icon" />}
                                            findNounPlural="samples"
                                            onSearch={onSearchIconClick}
                                            onFindByIds={onFindByIds}
                                        />
                                    ) : (
                                        <i
                                            className="fa fa-search navbar__xs-search-icon"
                                            onClick={onSearchIconClick}
                                        />
                                    )}
                                </div>
                            </div>
                        )}
                    </div>
                </div>
            </div>
        </nav>
    );
});

NavigationBar.defaultProps = {
    showNavMenu: true,
    showNotifications: true,
    showProductNav: true,
    showSearchBox: false,
};

NavigationBar.displayName = 'NavigationBar';
