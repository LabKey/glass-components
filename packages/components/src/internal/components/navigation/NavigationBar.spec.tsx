/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import renderer from 'react-test-renderer';
import { mount, ReactWrapper } from 'enzyme';
import { List } from 'immutable';

import { TEST_USER_GUEST, TEST_USER_READER } from '../../../test/data/users';
import { ServerNotifications } from '../notifications/ServerNotifications';

import { markAllNotificationsRead } from '../../../test/data/notificationData';
import { ServerNotificationModel } from '../notifications/model';

import { MenuSectionModel, ProductMenuModel } from './model';

import { NavigationBar } from './NavigationBar';
import { UserMenu } from "./UserMenu";
import { ProductMenu } from "./ProductMenu";
import { SearchBox } from "./SearchBox";
import { ProductNavigation } from "../productnavigation/ProductNavigation";
import moment from "moment";


beforeEach(() => {
    LABKEY.devMode = false;
    LABKEY.moduleContext = { samplemanagement: { hasPremiumModule: false } };
});

describe('<NavigationBar/>', () => {
    const productMenuModel = new ProductMenuModel({
        productIds: ['testNavBar'],
        isLoaded: true,
        isLoading: false,
        sections: List<MenuSectionModel>(),
    });

    const notificationsConfig = {
        maxRows: 1,
        markAllNotificationsRead,
        serverActivity: new ServerNotificationModel(),
        onViewAll: jest.fn(),
    };

    test('default props', () => {
        const component = <NavigationBar model={null} />;

        const tree = renderer.create(component);
        expect(tree).toMatchSnapshot();
    });

    test('with search box', () => {
        const component = <NavigationBar model={null} showSearchBox={true} />;

        const tree = renderer.create(component);
        expect(tree).toMatchSnapshot();
    });

    function validate(wrapper: ReactWrapper, compCounts?: Record<string, number>) {
        expect(wrapper.find('.project-name')).toHaveLength(compCounts?.ProjectName ?? 0);
        expect(wrapper.find(ProductMenu)).toHaveLength(compCounts?.ProductMenu ?? 0);
        expect(wrapper.find(UserMenu)).toHaveLength(compCounts?.UserMenu ?? 0);
        expect(wrapper.find(SearchBox)).toHaveLength(compCounts?.SearchBox ?? 0);
        expect(wrapper.find('.navbar__xs-search-icon')).toHaveLength(compCounts?.SearchBox ?? 0);
        expect(wrapper.find(ServerNotifications)).toHaveLength(compCounts?.ServerNotifications ?? 0);
        expect(wrapper.find(ProductNavigation)).toHaveLength(compCounts?.ProductNavigation ?? 0);
    }

    test('with notifications no user', () => {
        const component = mount(<NavigationBar model={productMenuModel} notificationsConfig={notificationsConfig} />);
        validate(component, { ProductMenu: 1, ServerNotifications: 0 });
        component.unmount();
    });

    test('with notifications, guest user', () => {
        const component = mount(
            <NavigationBar model={productMenuModel} user={TEST_USER_GUEST} notificationsConfig={notificationsConfig} />
        );
        validate(component, { ProductMenu: 1, UserMenu: 1, ServerNotifications: 0 });
        component.unmount();
    });

    test('with notifications, non-guest user', () => {
        const component = mount(
            <NavigationBar model={productMenuModel} user={TEST_USER_READER} notificationsConfig={notificationsConfig} />
        );
        validate(component, { ProductMenu: 1, UserMenu: 1, ServerNotifications: 1 });
        component.unmount();
    });

    test('show ProductNavigation for devMode', () => {
        LABKEY.devMode = true;
        const component = mount(
            <NavigationBar model={null} />
        );
        validate(component, { ProductNavigation: 1 });
        component.unmount();
    });

    test('show ProductNavigation for hasPremiumModule', () => {
        LABKEY.moduleContext = { samplemanagement: { hasPremiumModule: true } };
        const component = mount(
            <NavigationBar model={null} />
        );
        validate(component, { ProductNavigation: 1 });
        component.unmount();
    });
});
