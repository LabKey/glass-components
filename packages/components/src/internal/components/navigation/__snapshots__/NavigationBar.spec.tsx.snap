// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavigationBar/> default props 1`] = `
<nav
  className="navbar navbar-container test-loc-nav-header"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="navbar-left col-sm-5 col-xs-7"
      >
        <span
          className="navbar-item pull-left"
        />
        <span
          className="navbar-item-padded"
        />
      </div>
      <div
        className="navbar-right col-sm-7 col-xs-5"
      >
        <div
          className="navbar-item pull-right"
        />
        <div
          className="navbar-item pull-right"
        />
        <div
          className="navbar-item pull-right navbar-item-notification"
        />
        <div
          className="navbar-item pull-right hidden-xs"
        />
        <div
          className="navbar-item pull-right visible-xs"
        />
      </div>
    </div>
  </div>
</nav>
`;

exports[`<NavigationBar/> with search box 1`] = `
<nav
  className="navbar navbar-container test-loc-nav-header"
>
  <div
    className="container"
  >
    <div
      className="row"
    >
      <div
        className="navbar-left col-sm-5 col-xs-7"
      >
        <span
          className="navbar-item pull-left"
        />
        <span
          className="navbar-item-padded"
        />
      </div>
      <div
        className="navbar-right col-sm-7 col-xs-5"
      >
        <div
          className="navbar-item pull-right"
        />
        <div
          className="navbar-item pull-right"
        />
        <div
          className="navbar-item pull-right navbar-item-notification"
        />
        <div
          className="navbar-item pull-right hidden-xs"
        >
          <form
            className="navbar__search-form"
            onSubmit={[Function]}
          >
            <div
              className="form-group"
            >
              <i
                className="fa fa-search navbar__search-icon"
              />
              <input
                className="navbar__search-input"
                onChange={[Function]}
                placeholder="Enter Search Terms"
                size={34}
                type="text"
                value=""
              />
            </div>
          </form>
        </div>
        <div
          className="navbar-item pull-right visible-xs"
        >
          <i
            className="fa fa-search navbar__xs-search-icon"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
  </div>
</nav>
`;
